x-pack/filebeat/input/httpjson/config_okta_auth.go

```diff
@@ -0,0 +1,192 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package httpjson
+
+import (
+	"context"
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+// customTokenSource is a custom implementation of the oauth2.TokenSource interface.
+// for more information, see https://pkg.go.dev/golang.org/x/oauth2#TokenSource
+type customTokenSource struct {
+	mu      sync.Mutex
+	ctx     context.Context
+	conf    *oauth2.Config
+	token   *oauth2.Token
+	oktaJWK string
+}
+
+// fetchOktaOauthClient fetches an OAuth2 client using the Okta JWK credentials.
+func (o *oAuth2Config) fetchOktaOauthClient(ctx context.Context, _ *http.Client) (*http.Client, error) {
+	oktaJWK := string(o.OktaJWKJSON)
+	conf := &oauth2.Config{
+		ClientID: o.ClientID,
+		Scopes:   o.Scopes,
+		Endpoint: oauth2.Endpoint{
+			TokenURL: o.TokenURL,
+		},
+	}
+
+	oktaJWT, err := generateOktaJWT(oktaJWK, conf)
+	if err != nil {
+		return nil, fmt.Errorf("oauth2 client: error generating Okta JWT: %w", err)
+	}
+
+	token, err := exchangeForBearerToken(ctx, oktaJWT, conf)
+	if err != nil {
+		return nil, fmt.Errorf("oauth2 client: error exchanging Okta JWT for bearer token: %w", err)
+	}
+
+	tokenSource := &customTokenSource{
+		conf:    conf,
+		ctx:     ctx,
+		oktaJWK: oktaJWK,
+		token: &oauth2.Token{
```

Comment:
Why does this create a new oauth2.Token instead of passing through the `token` returned by `exchangeForBearerToken`? This might be worthy of a comment if there is a reason. e.g.

```
	token: token,
```

My concern is that the explicit `ExpiryTime` could be lost.

Reply:
Yes this was redundant. Will pass the token directly